compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -1341,41 +1363,47 @@ object desugar {
         }
       }
 
-      def isIrrefutableGenFrom(gen: GenFrom): Boolean =
-        gen.isInstanceOf[IrrefutableGenFrom] ||
-        IdPattern.unapply(gen.pat).isDefined ||
-        isIrrefutable(gen.pat, gen.expr)
+      def needsNoFilter(gen: GenFrom): Boolean =
+        if (gen.checkMode == GenCheckMode.FilterAlways) // pattern was prefixed by `case`
+          isIrrefutable(gen.pat, gen.expr)
```

Comment:
As a matter of principle, I think that if the user explicitly requested a "filtering binding" by writing `case`, then it is less surprising if we always emit a `withFilter` call. With `zipped`, we've already seen that emitting or not emitting the call can have unobvious consequences on typechecking. I think it'd be for the best if we just let the user decide whether the call is emitted or not.

Reply:
Good point. I'll change it.